oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/PersistenceTimer.java

```diff
@@ -92,73 +105,165 @@ public void start(ModuleManager moduleManager, CoreModuleConfig moduleConfig) {
-                    executorService.submit(() -> {
+            Future<?> batchFuture = batchExecutorService.submit(() -> {
+                List<Future<?>> results = new ArrayList<>();
+                // consume the metrics
+                while (!stop.get()) {
+                    List<PrepareRequest> partition = prepareQueue.popMany();
+                    if (partition.isEmpty()) {
+                        break;
+                    }
+                    Future<?> submit = executorService.submit(() -> {
+                        HistogramMetrics.Timer executeLatencyTimer = executeLatency.createTimer();
                         try {
                             if (CollectionUtils.isNotEmpty(partition)) {
                                 batchDAO.synchronous(partition);
                             }
                         } catch (Throwable e) {
                             log.error(e.getMessage(), e);
                         } finally {
-                            countDownLatch.countDown();
+                            executeLatencyTimer.finish();
                         }
+
                     });
+                    results.add(submit);
                 }
-                countDownLatch.await();
-            } finally {
-                executeLatencyTimer.finish();
-            }
+
+                if (!stop.get()) {
+                    for (Future<?> result : results) {
+                        result.get();
+                    }
+                }
+                return null;
+            });
+
+            // Wait for prepare stage is done.
+            prepareStageCountDownLatch.await();
+            prepareQueue.disableNeedFillFully();
+            // Wait for batch stage is done.
+            batchFuture.get();
+
         } catch (Throwable e) {
             errorCounter.inc();
             log.error(e.getMessage(), e);
         } finally {
+
             if (log.isDebugEnabled()) {
                 log.debug("Persistence data save finish");
             }
 
-            prepareRequests.clear();
+            stop.set(true);
+            allTimer.finish();
             lastTime = System.currentTimeMillis();
         }
 
         if (debug) {
             log.info("Batch persistence duration: {} ms", System.currentTimeMillis() - startTime);
         }
     }
+
+    private static class BlockingBatchQueue<E> {
+
+        private int maxBatchSize;
+        private boolean inAppendingMode = true;
+
+        public BlockingBatchQueue(int maxBatchSize) {
+            this.maxBatchSize = maxBatchSize;
+        }
+
+        private final List<E> elementData = new ArrayList<>(50000);
+
+        public void putMany(List<E> elements) {
+            synchronized (elementData) {
+                elementData.addAll(elements);
+                if (elementData.size() >= maxBatchSize) {
+                    elementData.notify();
+                }
+            }
+        }
```

Comment:
should there be a queue full check?
personally, I don't recommend implementing a data structure like this in so important module.

Reply:
The `private final List<PrepareRequest> prepareRequests = new ArrayList<>(50000);` did not check queue full before.
I think that there shouldn't be a queue full check.

Reply:
This is an Array list, it could resize. What do you mean full check? @dmsolr